packages/ckeditor5-dev-env/lib/release-tools/tasks/bump-year.js

```diff
@@ -0,0 +1,73 @@
+/**
+ * @license Copyright (c) 2003-2021, CKSource - Frederico Knabben. All rights reserved.
+ * For licensing, see LICENSE.md.
+ */
+
+'use strict';
+
+const fs = require( 'fs' );
+const glob = require( 'glob' );
+const minimatch = require( 'minimatch' );
+
+/**
+ * See https://github.com/ckeditor/ckeditor5/issues/10909
+ *
+ * @param {Object} params
+ * @param {String} params.cwd Current working directory from which all paths will be resolved
+ * @param {Function} params.callback Optional function to modify the license header
+ * @param {Number} params.initialYear Year from which the licenses should begin
+ * @param {Array} params.globPatterns array of objects, where each object has string property 'pattern',
+ * and optionally 'options' property.
+ */
+module.exports = function bumpYear( params ) {
+	params.globPatterns
+		.map( globPattern => glob.sync( globPattern.pattern, globPattern.options ) )
+		.reduce( ( previous, current ) => [ ...previous, ...current ] )
+		.filter( fileName => {
+			// Filter out nested `node_modules`.
+			if ( minimatch( fileName, '**/node_modules/**' ) ) {
+				return false;
+			}
+
+			// Filter out stuff from `src/lib/`.
+			if ( minimatch( fileName, '**/src/lib/**' ) ) {
+				return false;
+			}
+
+			// Filter out builds.
+			if ( minimatch( fileName, '**/ckeditor5-build-*/**' ) ||
+				minimatch( fileName, '**/build/**' ) ) {
+				return false;
+			}
+
+			// Filter out directories.
+			if ( fs.statSync( fileName ).isDirectory() ) {
+				return false;
+			}
+
+			return true;
+		} )
+		.forEach( fileName => {
+			const currentYear = new Date().getFullYear();
+			const yearOnlyRegexp = /(?<=^\/\*\*$[\s\S]+^ \* @license.+)(\d{4}-\d{4})(?=[\s\S]+^ \*\/$)/m;
+			const fullLicenseRegexp = /^\/\*\*$[\s\S]+^ \* @license.+\d{4}-\d{4}[\s\S]+^ \*\/$/m;
```

Comment:
Would it be possible to use only one regular expression to match the whole license string and then the year would be just a matched group inside it? Having 2 regular expressions seems like unnecessary duplication. Also it looks like the regular expression pattern is too generic, i.e. it could include the `Copyright (c)` pattern as well (which is AFAIK present in all the files), so there would be less false matches returned.

Reply:
The problem i had was, how to replace only the year, while also matching the entire regexp? Solution i found, was to use positive lookahead and lookbehind in order to find whole license, but only replace the part i want. So we look for the entire license, but only match the year, which we can then easily replace: `(?<=...)(2003-2021)(?=...)`. If you know of a better way to match whole license, split it in to 3 groups _(prefix / year / suffix),_ use one `.replace` function to replace only the year, and then use another `.replace` to modify the entire thing, then we can implement it. This solution produces two regexes, but results in much cleaner `.replace` functions. (Other solutions i was tinkering with ended up with much messier `.replace`, and were hard to make them work.)

Reply:
If the answer for [this question](https://github.com/ckeditor/ckeditor5/issues/10909#issuecomment-996748334) is that we should unify the license globally, then there is even a simpler solution: define the default license on top of this file to something like:

```js
const DEFAULT_LICENSE = `
/**
 * @license Copyright (c) %START_YEAR%-%END_YEAR%, CKSource Holding sp. z o.o. All rights reserved.
 * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-oss-license
 */
`;
```

Then, it is sufficient to just replace the `%START_YEAR%` and `%END_YEAR%` placeholders with appropriate values.
IMHO there is no need to use regular expressions to search for the years in the license content.

Reply:
This does not account for different file formats that have different syntax for comments, since we need to replace the license year everywhere.

[https://github.com/ckeditor/ckeditor5/blob/master/LICENSE.md?plain=1#L5](https://github.com/ckeditor/ckeditor5/blob/master/LICENSE.md?plain=1#L5)

This line does not even have `@license` prefix, so we either

1.  replace only `Copyright (c) 2003-2021`
2.  write a whole lot of different cases to cover all of the possibilities